File: day_16_JS/Advent_16_1.js
"use strict"

const fs = require('fs');

let file_data;

const file = "dummy.txt"
// const file = "data.txt"


fs.readFile(file, 'utf8', (err, data) => {
    if(err) throw err;
    file_data = data.split("\r\n")

    const pipes = {}

    for(let line = 0; line < file_data.length; line++){
        let name = file_data[line].split(" ")[1];
        let rate = parseInt(file_data[line].split(" ")[4].slice(5, -1));
        pipes[name] = new Pipe(name, rate)
    }
    for(let line = 0; line < file_data.length; line++){
        let name = file_data[line].split(" ")[1];
        let raw_outs = file_data[line].split(" ").slice(9)
        for(const out of raw_outs){
            pipes[name].outs.push(pipes[out.slice(0,2)])
        }
    }

    const trip = new Trip(pipes, "AA")

    // print(trip.max_value(30))

    // print(pipes["DD"].total_value(pipes, 28))
    // pipes["DD"].open()
    // print(pipes["AA"].total_value(pipes, 26))
    // print(pipes["BB"].total_value(pipes, 25))
    pipes["AA"].next_node(pipes, 30)
    // pipes["DD"].open()
    // pipes["DD"].next_node(pipes, 28)
    // print(pipes["EE"].total_value(pipes, 26))
    print(pipes["DD"].value(pipes["EE"]), 28)
    
});

class Pipe {
    constructor(name, rate){
        this.name = name;
        this.rate = rate;
        this.outs = [];
        this.is_open = false;
        this.released = 0;
    }
    
    open(){
        this.is_open = true;
    }
    
    step(n=1){
        if(this.is_open){this.released += this.rate * n}
    }
    
    path(target, open=new Set([this]), closed=new Set()){
        let running = true;
        let counter = 0;
        if(target == this.name){
            return null
        }
        
        while(running){
            const holder = []
            counter += 1;
            for(const node of open){
                closed.add(node.name)
                open.delete(node)
                for(const child of node.outs){
                    if(!(child.name in closed)){
                        if(child.name == target){
                            running = false;
                        }
                        holder.push(child)
                    }
                }
            }
            
            for(const child of holder){
                open.add(child)
            }
        }
        // print(`Found ${target} in ${counter} steps.`)
        return counter
    }
    
    value(target_node, time_left=30){
        const distance = this.path(target_node.name)
        
        if(distance == null){
            return {name: target_node.name, distance: 0, value: target_node.rate * time_left, open: target_node.is_open}
        }
        let value = 0;
        if(!target_node.is_open){
            time_left -= distance + 1
            value += time_left * target_node.rate
        }
        
        return { name: target_node.name, distance: distance, value, open: target_node.is_open}
    }
    
    total_value(pipes, time_left=30){
        let sum = 0;
        for(const name of Object.keys(pipes)){
            if(!pipes[name].is_open){
                const result = pipes[this.name].value(pipes[name], time_left)
                sum += result.value
            }
        }
        return sum
    }
    
    next_node(pipes, time_left=30){
        let max = {pipe: null, value: -1};
        for(const name of Object.keys(pipes).filter((name) => name != this.name)){
            const distance = this.path(pipes[name].name)
            const result = pipes[name].total_value(pipes, time_left - distance-1)
            print(name, result)
            if(result > max.value && name != this.name && pipes[name].rate > 0 && !pipes[name].is_open){
                max = {pipe: pipes[name], value: result}
            } else {
                // print(name, result)
            }
        }
        return max
    }
}

class Trip {
    constructor(graph, head){
        this.graph = graph
        this.head = graph[head]
        this.total_released = 0; 
        this.total_rate = 0;
        this.total_opened = 0; 
    }
    
    max_value(time_left=30){
        while (time_left > 1){
            let next_node = this.head.next_node(this.graph, time_left).pipe
            print(this.head.name)
            print(this.head.is_open)
            let time_elapsed = this.head.path(next_node.name) + 1
            time_left -= time_elapsed
            this.total_released += this.total_rate * time_elapsed
            next_node.open()
            this.total_rate += next_node.rate
            this.total_opened += 1
            this.head = next_node
        }

        print(`Loop break with ${this.time_left} minutes left.`)
        return this.total_released
        
    }
    
    
}


function print(...x){
    console.log(...x)
}

File: day_16_ruby/Advent_16_1.rb
require "byebug"
require "set"
class PipeGraph
    attr_accessor :head, :total_released, :total_rate, :time_remaining, :order, :pipes, :name, :order
    
    def initialize(file_name, head)
        @total_released = 0
        @total_rate = 0
        @pipes = Hash.new()
        @time_remaining = 30
        @order = []
        @name = "#{self.object_id}"
        file = File.open(file_name)
        
        file_data = file.read()

        file_data = file_data.split("\r\n")

        file_data.each do |line|
            name = line.split(" ")[1]
            rate = line.split(" ")[4][5..-1].to_i
            Pipe.new(name, rate, self)
        end

        file_data.each do |line|
            name = line.split(" ")[1]
            raw_outs = line.split(" ")[9..-1]
            # p line
            # p raw_outs
            for out in raw_outs
                @pipes[name].outs << @pipes[out.slice(0, 2)]
            end
        end

        @head = @pipes[head]
    end
    
    def report
        puts "Time Remaining: #{@time_remaining}"
        puts "Total Rate: #{@total_rate}"
        puts "Total Released: #{@total_released}"
        # print "Order: #{@order}"
    end

    def run
        while @pipes.any?{|key, pipe| !pipe.open? && pipe.rate > 0} && @time_remaining > 0
            target = @pipes[self.choose_target()]

            if target
                @head.distance(target).times do
                    self.step(target.name)
                end
                self.open()
            else
                self.step()
            end
        end

        while @time_remaining > 0
            self.step()
        end
        self.report()
    end

    def choose_target


        holder = {}
        @pipes.each_value do |pipe|
            holder[pipe.name] = self.delta(pipe) unless @head == pipe || pipe.rate < 1 || pipe.open? || @head.distance(pipe)+1 > @time_remaining
        end

        

        values = holder.values.sort
        running = true

        while running
            return nil unless values[0]
            candidate = @pipes[holder.key(values.shift)]

            @pipes.values.reject{|pipe| pipe == candidate || pipe.open?}.each do |pipe|
                running = false
                if (candidate.distance(pipe)*2 + 1) * candidate.rate < pipe.rate
                    running = true
                    break
                end
            end
        end
        return candidate.name
    end
    
    def delta(target_node)
        total = 0
        @pipes.each_value do |pipe|
            if pipe == target_node
                next
            end
            total += @head.value(pipe) unless pipe.open?
        end

        new_total = 0
        remaining = @time_remaining - @head.distance(target_node) - 1

        @pipes.each_value do |pipe|
            if pipe == target_node
                next
            end
            new_total += target_node.value(pipe, false, remaining) unless pipe.open?
        end

        delta = total - new_total

        if !@head.open?
            self_value = remaining * target_node.rate
        else
            self_value = 0
        end

        return delta - self_value

        # if self_value - delta < 0
        #     return Float::INFINITY
        # else
        #     return delta
        # end

    end

    def value(target_node)
        distance = @head.distance(target_node)
        elapsed = distance + 1
        # @pipes.each_value do |pipe|
        #     puts "#{pipe.name} #{target_node.value(pipe, true, @time_remaining - elapsed)}"
        # end
        if target_node.rate > 0 && !target_node.open?
            values = @pipes.values.map{|pipe| 
                target_node.value(pipe, true, @time_remaining - elapsed)
            }
            values << target_node.rate * (@time_remaining - elapsed)
        else
            values = [0]
        end

        puts "Total of #{target_node.name}: #{values.sum}"
    end

    def open()
        unless @head.open? || @time_remaining < 1
            self.step()
            @head.open!()
            @order << "Open #{@head.name}"
        end
    end

    def inspect
        return "#{self.class} #{self.name}"
    end

    def step(next_node_name = nil)
        if @time_remaining < 1
            puts "Time Over"
            return
        end
        @time_remaining -= 1
        @pipes.each_value {|pipe| pipe.step()}
        if next_node_name
            @head = @pipes[next_node_name]
            @order << "Step to #{next_node_name}"
        end
    end
end

class Pipe < PipeGraph
    attr_reader :open, :name, :outs, :released, :rate
    def initialize(name, rate, graph)
        @name = name
        @rate = rate
        @released = 0
        @open = false
        @outs = []
        @graph = graph
        graph.pipes[name] = self
    end

    def open?
        @open
    end

    def distance(target_node)
        on_deck = Set.new([self])
        visited = Set.new
        counter = 0
        running = true

        if target_node == self
            return 0
        end

        while running
            holder = []
            counter += 1
            for node in on_deck
                visited.add(node)
                on_deck.delete(node)
                for child in node.outs
                    if !visited.include?(child)
                        if child == target_node
                            running = false;
                        end
                        holder.push(child)
                    end
                end
            end
            for child in holder
                on_deck.add(child)
            end
        end
        # puts "Found #{target_node.name} in #{counter} steps."
        return counter
    end

    def value(target_node, mock_open = false, time_remaining = @graph.time_remaining)
        if target_node.open? || target_node.rate == 0 || (mock_open && target_node == self)
            return 0
        else
            distance = self.distance(target_node)
            # rate_holder = @graph.total_rate + (mock_open ? self.rate : 0)
            return (time_remaining - distance-1) * target_node.rate #+ (distance+1) * rate_holder
            
        end
    end

    def step()
        if @open
            @released += @rate
            @graph.total_released += @rate
        end
    end

    def open!()
        unless @open
            @open = true
            @graph.total_rate += @rate
        end
    end

end

puts
puts "TEST"
a = PipeGraph.new("dummy.txt", "AA")
a.run
puts

puts "PUZZLE"
a = PipeGraph.new("data.txt", "NQ")
a.run
puts